operator/pkg/ciliumendpointbatch/manager.go

```diff
@@ -0,0 +1,422 @@
+func randomName(n int) string {
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = sequentialLetters[rand.Intn(len(sequentialLetters))]
+	}
+	return string(b)
+}
+
+// Generates unique random name for the CiliumEndpointBatch, the format
+// of a CEB name is similar to pod k8s naming convention "ceb-123456789-abcde".
+// First 3 letters indicates ceb resource, followed by random letters.
+func uniqueCeBatchName(cacheCeb map[string]*ceBatch) string {
+	rand.Seed(time.Now().UnixNano())
+	var exists bool
+	var cebName string
+
+	for {
+		cebName = fmt.Sprintf("%s-%s-%s", cebNamePrefix, randomName(9), randomName(5))
+		if _, exists = cacheCeb[cebName]; !exists {
+			return cebName
+		}
+	}
+}
+
+// createCeb function create a new ceb and capacity to hold maximum ceps in a CEB.
+// This function called on 2 different scenarios.
+// 1) During runtime, when ceb manager decides to create a new ceb, it calls
+//    with an empty name. function generates random unique name and assign it to CEB
+// 2) During operator warm boot [after crash or software upgrade], batching manager
+//    create CEB, by passing unique name.
+func (c *cebManagerFcfs) createCeb(name string) *ceBatch {
+	var cebName string = name
+	if name == "" {
+		cebName = uniqueCeBatchName(c.cacheCeb)
+	}
+	ceb := &ceBatch{
+		ceb: &cilium_v2.CiliumEndpointBatch{
+			TypeMeta: meta_v1.TypeMeta{
+				Kind:       "CiliumEndpointBatch",
+				APIVersion: cilium_v2.SchemeGroupVersion.String(),
+			},
+			ObjectMeta: meta_v1.ObjectMeta{
+				Name: cebName,
+			},
+			Endpoints: make([]cilium_v2.CoreCiliumEndpoint, 0, c.maxCepsInCeb),
+		},
+		removedCeps: make(map[string]struct{}),
+	}
+	c.cacheCeb[cebName] = ceb
+	log.WithFields(logrus.Fields{
+		"ceb-name": cebName,
+	}).Debug("Generated ceb name")
+	return ceb
+}
+
+// If exists, remove Ceb object from cache. deleteCebFromCache called after successful removal from
+// apiserver.
+func (c *cebManagerFcfs) deleteCebFromCache(cebName string) {
+	if _, ok := c.cacheCeb[cebName]; !ok {
+		log.WithFields(logrus.Fields{
+			"ceb-name": cebName,
+		}).Debug("Failed to retrieve Ceb object in local cache.")
+		return
+	}
+	delete(c.cacheCeb, cebName)
+}
+
+// updateCebInCache function copies the ciliumEndpoint object in local cache. if isDeepCopy flag is set,
+// whole CoreCiliumEndpoint object stored in local cache.
+// There are two scenarios updateCebInCache is called.
+// 1) During operator warm boot[after crash or software upgrade], CEB controller sync CEB states from
+// api-server to cache. In this case, isDeepCopy set to true to copy entire CEP object locally.
+// 2) During runtime, reconciler sync current state with API server and update metadata only.
+// isDeepCopy flag is set to false.
+func (c *cebManagerFcfs) updateCebInCache(srcCeb *cilium_v2.CiliumEndpointBatch, isDeepCopy bool) {
+	if ceb, ok := c.cacheCeb[srcCeb.GetName()]; ok {
+		if !isDeepCopy {
+			ceb.ceb.ObjectMeta = srcCeb.ObjectMeta
+		} else {
+			srcCeb.DeepCopyInto(ceb.ceb)
```

Comment:
This will deep copy fields form `srcCeb` only if they are set. Which means that if you have a `ceb.ceb` that have `field A` and `field B` set but `srcCeb` only has `field A` set, the end result will be a `ceb.ceb` with `srcCeb`'s `field A` and the old `ceb.ceb`'s field B.

Reply:
As i mentioned in the comment for the function `updateCebInCache`, `deepCopy`ing of `ceb` is only done at operator warm boot.  When the operator comes up, it needs to sync current the CEB states from `api-server` to local cache. Hence we deepCopy, at runtime we never do `deepCopy` of entire CEB.

Reply:
I think it's worthing separating this function into two for better readability/code structure: 
- one that only does "updateMeta", aka without deep copy, this one is only called to maintain "freshness" of local copy
- one that does deep copy, and is only called during operator reboot (name should indicate that, like "populateCacheFromWatchStore"

Reply:
I don't see any need  for separate into two functions. The current function suits well here.

Reply:
Well, i don't see that way.  Again this function is well commented and use cases under what scenarios this called are explained well.

Reply:
Since we cannot reach consensus, @varunmar PTAL

Reply:
Agree with your analysis, this function gets called during operator boot up time. We have no `objects` in `ceb.ceb` except the fact it was pre built with some basic information.
Anyways, I'll address your concern.